### ISSUE TITLE
Behavior of mixed derivatives is unexpected

### ISSUE THREAD

**Issue by jotterbach** [pytorch/pytorch#33999]
## 🐛 Bug

For a research project, I want to implement a mixed second derivative wrt. to the input *and* parameters, i.e. $\partial_x\partial_\theta L(x, \theta)$. This works reasonably well using the `torch.autograd.grad` functionality as proven in the snippet below, but incurs significant memory overhead due to having to store a separate set of activations (which I am not sure why yet) and prohibits the use of `torch.utils.checkpoint` functionality if the mixed derivative happens in the checkpointing graph. Also see #24005.

On the other hand I can use the `torch.autograd.backward` functionality to compute the derivative, leading to significantly smaller memory requirements and compatibility with `torch.utils.checkpoint`, but it turns out that the parameter gradients in this case are extremely wrong.

A small script highlighting the issue is attached below.

## To Reproduce

```python
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.utils.checkpoint as cp

seed = 142


class Square(nn.Module):

    def __init__(self):
        super(Square, self).__init__()
        self.a = nn.Parameter(torch.randn(1))

    def forward(self, x):
        return (self.a * x)**2


def grad_of_torch_func_backward(z, func):
    with torch.enable_grad():
        z.requires_grad_(True)
        omega = func(z).reshape(z.size(0), 1).sum()
        torch.autograd.backward(omega)
        dz = z.grad
        dz.requires_grad_(True)
        return dz


def grad_of_torch_func(z, func):
    with torch.enable_grad():
        z.requires_grad_(True)
        omega = func(z).reshape(z.size(0), 1)
        go = torch.ones_like(omega).to(z.device)
        dz = torch.autograd.grad(outputs=omega,
                                 inputs=z,
                                 grad_outputs=go,
                                 create_graph=True,
                                 retain_graph=True)[0]
        return dz


def objective_wrong_grad(model: nn.Module, y: torch.Tensor):
    y.requires_grad_(True)
    g = grad_of_torch_func_backward(y, model)
    y.requires_grad_(False)
    return y - g


def objective_correct_grad(model: nn.Module, y: torch.Tensor):
    y.requires_grad_(True)
    g = grad_of_torch_func(y, model)
    y.requires_grad_(False)
    return y - g


def correct_param_autograd():
    torch.manual_seed(seed)
    se = Square()

    x = torch.ones(10, 1)
    y = x + 0.1 * torch.randn_like(x)
    z = objective_correct_grad(se, y)

    loss = cp.checkpoint(F.mse_loss, z, x)
    loss.backward()

    param_grad_correct = se.a.grad.item()
    param = se.a.item()

    return x.data.numpy(), y.data.numpy(), z.data.numpy(), loss.item(), param, param_grad_correct


def incorrect_param_autograd():
    torch.manual_seed(seed)
    se = Square()

    x = torch.ones(10, 1)
    y = x + 0.1 * torch.randn_like(x)
    z = objective_wrong_grad(se, y)

    loss = F.mse_loss(z, x)
    loss.backward()

    param_grad_incorrect = se.a.grad.item()
    param = se.a.item()

    return x.data.numpy(), y.data.numpy(), z.data.numpy(), loss.item(), param, param_grad_incorrect


def compute_analytic_param_grad(a, x, y):
    return (8 * a * y * (x + (2 * a**2 - 1) * y)).mean().item()


def main():
    x0, y0, z0, loss0, p0, p_grad0 = correct_param_autograd()
    x1, y1, z1, loss1, p1, p_grad1 = incorrect_param_autograd()

    np.testing.assert_almost_equal(p0, p1)
    np.testing.assert_array_almost_equal(x0, x1)
    np.testing.assert_array_almost_equal(y0, y1)
    np.testing.assert_array_almost_equal(z0, z1)
    np.testing.assert_almost_equal(loss0, loss1)

    analytic_grad = compute_analytic_param_grad(p0, x0, y0)
    np.testing.assert_almost_equal(p_grad0, analytic_grad)

    np.testing.assert_almost_equal(p_grad0, p_grad1)


if __name__ == '__main__':
    main()
```

## Expected behavior

Enable the ability to compute correct gradient wrt to the inputs using `grad` in conjunction with checkpointing in order to be able to scale to larger models. Alternatively enable correct mixed derivatives with `backward` such that the memory overhead is reduced and checkpointing can be used.

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.1
GCC version: Could not collect
CMake version: version 3.15.2

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] pytorch-memlab==0.0.4
[pip3] torch==1.3.1
[pip3] torchtext==0.4.0
[pip3] torchvision==0.4.2
[pip3] torchviz==0.0.1
[conda] Could not collect

## Additional context

See #24005

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by albanD** [comment]
Hi,

Thanks for the runnable code sample!

few things:
- You should not need to set/unset the requires_grad field this way:
  - For the parameters you learn, the `nn.Parameter` takes care of it for you.
  - For your custom functions, you only need it if it not already set to create a new leaf that will require gradients. 
  - You should never have to set it to False
- If you set `create_graph=True`, then `retain_graph` will be true by default, no need to set it
- `grad_of_torch_func_backward` is not setting `create_graph=True` when calling backward, I think it should
- Both your functions don't reset the gradients of the module before calling the final loss.backward(). In this case, the `.backward()` call that you did earlier already set some values in there, so it should be reset if you only want the gradient from the loss.

After doing these changes, your code run fine for me.
Feel free to re-open if you still see problems.


Note, for questions, we are usually more responsive if you ask on the forum: https://discuss.pytorch.org/ as more people will be looking at your code.
